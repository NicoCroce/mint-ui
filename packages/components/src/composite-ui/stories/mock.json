{
    "type": "help_center",
    "children": [
        {
            "type": "header"
        },
        {
            "type": "section_title"
        },
        {
            "type": "list",
            "children": [
                {
                    "type": "list_item",
                    "title": "Insurance",
                    "expandRight": true,
                    "url": "/help-center/insurance"
                },
                {
                    "type": "list_item",
                    "title": "Mutual Funds",
                    "expandRight": true,
                    "url": "/help-center/mutual-funds"
                },
                {
                    "type": "list_item",
                    "title": "Gold",
                    "expandRight": true,
                    "url": "/help-center/gold"
                },
                {
                    "type": "list_item",
                    "title": "Loans",
                    "expandRight": true,
                    "url": "/help-center/loans"
                },
                {
                    "type": "list_item",
                    "title": "Something Else",
                    "expandRight": true,
                    "url": "/help-center/something-else"
                }
            ]
        }
    ]
}