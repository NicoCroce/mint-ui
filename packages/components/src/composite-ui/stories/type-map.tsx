import { ListTypeWrapper } from "./type-wrappers/list";
import { Wrapper } from "./type-wrappers/wrapper";
import { renderComponent } from "..";
import { IDict } from "../../__utils/type";
import { ListItem } from "../../list/list-item";

// Search on how to type any part of it fully since it return an exact signature
export const TypeMap: IDict<(Component: any) => any> = {
	list: renderComponent(ListTypeWrapper),
	list_item: renderComponent(ListItem),
	wrapper: renderComponent(Wrapper)
};
