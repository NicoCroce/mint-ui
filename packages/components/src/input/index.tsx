import * as React from "react";
import styled, { css } from "styled-components";
import { Omit, CommonTypeTuple } from "../__utils/type";
import COLORS from "../__utils/colors";
import { GlobalStyles } from "../app";

export interface InputProps {
	type?: string;
	value?: CommonTypeTuple;
	placeholder?: string;
	label?: string;
	error?: string;
	helpText?: string;
	block?: boolean;
	disabled?: boolean;
	onChange?: (val: string, name: string) => void;
	onBlur?: React.FocusEventHandler;
}

export const InputStyles = css<
	Pick<InputProps, "error" | "disabled"> & Omit<InputProps, "onChange">
>`
	box-sizing: border-box;
	border: 0;
	border: 1px solid ${COLORS.GREY4};
	outline: none;
	padding: 12px 48px 12px 16px;
	border-radius: 4px;
	color: ${COLORS.GREY1};
	width: 100%;
	font-size: 14px;
	&:focus {
		border-color: ${COLORS.PRIMARY};
	}
	::-webkit-input-placeholder {
		/* Chrome/Opera/Safari */
		color: ${COLORS.DISABLED};
	}
	::-moz-placeholder {
		/* Firefox 19+ */
		color: ${COLORS.DISABLED};
	}
	:-ms-input-placeholder {
		/* IE 10+ */
		color: ${COLORS.DISABLED};
	}
	:-moz-placeholder {
		/* Firefox 18- */
		color: ${COLORS.DISABLED};
	}
	::placeholder: {
		color: ${COLORS.DISABLED};
	}
	${props =>
		props.error &&
		css`
			border-color: ${COLORS.DANGER};
			&:focus {
				border-color: ${COLORS.DANGER};
			}
		`};
	${props =>
		props.disabled &&
		css`
			opacity: 0.64;
			cursor: no-drop;
			background-color: ${COLORS.DISABLED};
			color: ${COLORS.GREY3};
			::-webkit-input-placeholder {
				/* Chrome/Opera/Safari */
				color: ${COLORS.GREY3};
			}
			::-moz-placeholder {
				/* Firefox 19+ */
				color: ${COLORS.GREY3};
			}
			:-ms-input-placeholder {
				/* IE 10+ */
				color: ${COLORS.GREY3};
			}
			:-moz-placeholder {
				/* Firefox 18- */
				color: ${COLORS.GREY3};
			}
			::placeholder: {
				color: ${COLORS.GREY3};
			}
		`};
`;

const SharedStyles = css`
	color: ${COLORS.GREY1};
`;

export const LabelStyles = css<{ error: string | undefined }>`
	margin-bottom: 4px;
	label {
		${SharedStyles};
		${({ error }) =>
			error &&
			css`
				color: ${COLORS.DANGER};
			`};
	}
`;

export const Error = css`
	${SharedStyles};
	margin-top: 4px;
	word-wrap: break-word;
	label {
		color: ${COLORS.DANGER};
	}
`;

export const HelpText = css`
	${SharedStyles};
	color: ${COLORS.GREY2};
`;

const StyledWrapper = styled.div<{ block: boolean }>`
	${GlobalStyles};
	width: ${({ block }) => (block ? "100%" : "328px")};
	text-align: left;
`;

const StyledLabel = styled.div<{ error: undefined | string }>`
	${LabelStyles};
	font-weight: 500;
	color: ${({ error }) => (error ? COLORS.DANGER : "initial")};
`;
export const StyledInput = styled.input<
	Pick<InputProps, "error" | "disabled"> & Omit<InputProps, "onChange">
>`
	${InputStyles};
`;
const StyledError = styled.div<{ children: React.ReactNode }>`
	${Error};
`;
const StyledHelpText = styled.p<{ children: React.ReactNode }>`
	${HelpText};
`;

export const Input: React.FC<InputProps &
	Omit<React.InputHTMLAttributes<HTMLInputElement>, "onChange">> = ({
	type = "text",
	placeholder,
	label,
	error,
	helpText,
	block = false,
	disabled = false,
	onChange,
	onBlur,
	value,
	className = "",
	...rest
}: InputProps & React.InputHTMLAttributes<HTMLInputElement>) => {
	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		if (onChange) {
			onChange(e.target.value, e.target.name);
		}
	};
	return (
		<StyledWrapper block={block}>
			{label ? (
				<StyledLabel error={error}>
					<label>{label}</label>
				</StyledLabel>
			) : null}
			<StyledInput
				type={type}
				placeholder={placeholder}
				error={error}
				disabled={disabled}
				onBlur={onBlur}
				onChange={handleChange}
				value={value}
				className={className}
				aria-label="tm-input"
				{...rest}
			/>
			{error ? (
				<StyledError>
					<label>{error}</label>
				</StyledError>
			) : null}
			{helpText ? <StyledHelpText>{helpText}</StyledHelpText> : null}
		</StyledWrapper>
	);
};

export default Input;
