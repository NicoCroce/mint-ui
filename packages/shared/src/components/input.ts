import { css } from "styled-components";
import { PALETTE } from "../utils/colors";
import { Omit, CommonTypeTuple } from "../utils/type";

export interface InputProps {
	type?: string;
	value: CommonTypeTuple;
	placeholder?: string;
	label?: string;
	error?: string;
	helpText?: string;
	block?: boolean;
	disabled?: boolean;
	onChange?: (val: string) => void;
	onBlur?: React.FocusEventHandler;
}

export const Wrapper = css<Pick<InputProps, "block">>`
	width: ${props => (props.block ? "100%" : "328px")};
	text-align: left;
`;

export const Input = css<
	Pick<InputProps, "error" | "disabled"> & Omit<InputProps, "onChange">
>`
	border: 0;
	border: 1px solid ${PALETTE.DISABLED};
	padding: 14px 48px 14px 16px;
	border-radius: 4px;
	color: ${PALETTE.GREY1};
	font-size: 16px;
	width: 100%;
	&:focus {
		border: 2px solid ${PALETTE.PRIMARY};
	}
	::-webkit-input-placeholder {
		/* Chrome/Opera/Safari */
		color: ${PALETTE.DISABLED};
	}
	::-moz-placeholder {
		/* Firefox 19+ */
		color: ${PALETTE.DISABLED};
	}
	:-ms-input-placeholder {
		/* IE 10+ */
		color: ${PALETTE.DISABLED};
	}
	:-moz-placeholder {
		/* Firefox 18- */
		color: ${PALETTE.DISABLED};
	}
	::placeholder: {
		color: ${PALETTE.DISABLED};
	}
	${props =>
		props.error &&
		css`
			border: 1px solid ${PALETTE.DANGER};
			&:focus {
				border: 1px solid ${PALETTE.DANGER};
			}
		`};
	${props =>
		props.disabled &&
		css`
			opacity: 0.64;
			cursor: no-drop;
			background-color: ${PALETTE.DISABLED};
			color: ${PALETTE.GREY3};
			::-webkit-input-placeholder {
				/* Chrome/Opera/Safari */
				color: ${PALETTE.GREY3};
			}
			::-moz-placeholder {
				/* Firefox 19+ */
				color: ${PALETTE.GREY3};
			}
			:-ms-input-placeholder {
				/* IE 10+ */
				color: ${PALETTE.GREY3};
			}
			:-moz-placeholder {
				/* Firefox 18- */
				color: ${PALETTE.GREY3};
			}
			::placeholder: {
				color: ${PALETTE.GREY3};
			}
		`};
`;

const Shared = css`
	color: ${PALETTE.GREY1};
	font-size: 14px;
	line-height: 1.43;
`;
type LabelProps = Pick<InputProps, "error">;

export const Label = css<LabelProps>`
	margin-bottom: 4px;
	label {
		${Shared};
		${props =>
			props.error &&
			css`
				color: ${PALETTE.DANGER};
			`};
	}
`;

export const Error = css`
	${Shared};
	margin-top: 4px;
	word-wrap: break-word;
	label {
		color: ${PALETTE.DANGER};
	}
`;

export const HelpText = css`
	${Shared};
	color: ${PALETTE.GREY2};
`;
