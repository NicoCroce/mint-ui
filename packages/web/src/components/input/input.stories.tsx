import * as React from "react";
import { storiesOf } from "@storybook/react";
import Row from "@turtlemint/tm-components-shared/src/components/row";
import Col from "@turtlemint/tm-components-shared/src/components/col";
import { Input } from "./index";

const stories = storiesOf("Input", module);

export const TestInput = () => {
	const [value, setValue] = React.useState("");
	const handleChange = (val: string) => {
		setValue(val);
	};
	return (
		<Input
			type="text"
			placeholder="Test input"
			value={value}
			onChange={handleChange}
		/>
	);
};
stories.add("Input", () => (
	<Row
		style={{
			padding: "0 30px"
		}}
	>
		<Col sm={6} md={6} lg={4}>
			<TestInput />
		</Col>
	</Row>
));
