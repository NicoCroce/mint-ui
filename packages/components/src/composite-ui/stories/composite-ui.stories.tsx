import * as React from "react";
import { storiesOf } from "@storybook/react";
import { CompositeUI } from "../index";
import config from "./mock.json";
import { TypeMap } from "./type-map";

const stories = storiesOf("Composite UI", module);

stories.add("default", () => {
	return <CompositeUI config={config} typeMap={TypeMap} />;
});
