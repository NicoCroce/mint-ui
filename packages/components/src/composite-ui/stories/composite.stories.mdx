import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { CompositeUI } from "../index";
import config from "./mock.json";
import { TypeMap } from "./type-map";

<Meta title="CompositeUI" component={CompositeUI} />

# Composite UI

With `CompositeUI` you can render any nested layout with specified structure in config structure
in a single line


<Preview>
  <Story name="basic">
   <CompositeUI config={config} typeMap={TypeMap} />
  </Story>
</Preview>

### Props Table

<Props of={CompositeUI} />

### Config structure

```
{
    "type": "help_center",
    "children": [
        {
            "type": "header"
        },
        {
            "type": "section_title"
        },
        {
            "type": "list",
            "children": [
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "warning",
                            "color": "blue"
                        },
                        "title": "Insurance"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "close"
                        },
                        "title": "Mutual Funds"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "settings",
                            "color": "grey"
                        },
                        "title": "Gold"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "notification_important",
                            "color": "red"
                        },
                        "title": "Loans"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "info"
                        },
                        "title": "Something Else"
                    }
                }
            ]
        }
    ]
}
```
