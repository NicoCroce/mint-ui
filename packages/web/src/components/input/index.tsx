import * as React from "react";
import styled from "styled-components";
import * as Styles from "@turtlemint/tm-components-shared/src/components/input";
import { Omit } from "@turtlemint/tm-components-shared/src/utils/type";

const StyledWrapper = styled.div`
	${Styles.Wrapper};
`;
type LabelProps = Pick<Styles.InputProps, "error"> & {
	children: React.ReactNode;
};

const StyledLabel = styled.div<LabelProps>`
	${Styles.Label};
`;
const StyledInput = styled.input<
	Pick<Styles.InputProps, "error" | "disabled"> &
		Omit<Styles.InputProps, "onChange">
>`
	${Styles.Input};
`;
const StyledError = styled.div<{ children: React.ReactNode }>`
	${Styles.Error};
`;
const StyledHelpText = styled.p<{ children: React.ReactNode }>`
	${Styles.HelpText};
`;
export const Input: React.FC<
	Styles.InputProps &
		Omit<React.InputHTMLAttributes<HTMLInputElement>, "onChange">
> = ({
	type = "text",
	placeholder,
	label,
	error,
	helpText,
	block = false,
	disabled = false,
	onChange,
	onBlur,
	value,
	className,
	...rest
}: Styles.InputProps & React.InputHTMLAttributes<HTMLInputElement>) => {
	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		if (onChange) {
			onChange(e.target.value);
		}
	};
	return (
		<StyledWrapper block={block}>
			<StyledLabel error={error}>
				<label>{label}</label>
			</StyledLabel>
			<StyledInput
				type={type}
				placeholder={placeholder}
				error={error}
				disabled={disabled}
				onBlur={onBlur}
				onChange={handleChange}
				value={value}
				className={className}
				aria-labelledby="tm-input"
				{...rest}
			/>
			{error ? (
				<StyledError>
					<label>{error}</label>
				</StyledError>
			) : null}
			{helpText ? <StyledHelpText>{helpText}</StyledHelpText> : null}
		</StyledWrapper>
	);
};
// Input.TextArea = TextArea;

export default Input;
