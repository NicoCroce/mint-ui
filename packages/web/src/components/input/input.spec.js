import React from "react";
import { fireEvent, render, cleanup } from "@testing-library/react";
import { TestInput } from "./input.stories";

afterEach(cleanup);

describe("Input tests", () => {
	const { asFragment, getByLabelText } = render(<TestInput />);

	test("renders correcty", () => {
		expect(asFragment()).toMatchSnapshot();
	});

	test("Handles text change", () => {
		const input = getByLabelText("tm-input");
		fireEvent.change(input, { target: { value: "New value" } });
		expect(input.value).toBe("New value");
	});
});
