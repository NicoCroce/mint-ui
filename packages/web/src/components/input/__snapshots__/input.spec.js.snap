// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests renders correcty 1`] = `
<DocumentFragment>
  <div
    class="input__StyledWrapper-sc-1vl8hlv-0 fSzACW"
  >
    <div
      class="input__StyledLabel-sc-1vl8hlv-1 fORrYV"
    >
      <label />
    </div>
    <input
      aria-labelledby="tm-input"
      class="input__StyledInput-sc-1vl8hlv-2 dbjOrN akjas"
      placeholder="Test input"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
